hibernate-core/src/main/java/org/hibernate/loader/ast/internal/AbstractMultiNaturalIdLoader.java

```diff
@@ -57,7 +60,18 @@ private <K> List<E> performUnorderedMultiLoad(K[] naturalIds, MultiNaturalIdLoad
 		);
 	}
 
-	protected abstract <K> List<E> unorderedMultiLoad(K[] naturalIds, SharedSessionContractImplementor session, LockOptions lockOptions);
+	protected <K> List<E> unorderedMultiLoad(K[] naturalIds, SharedSessionContractImplementor session, LockOptions lockOptions) {
+		final List<E> results = arrayList( naturalIds.length );
+		final Object[] unresolvedIds =
+				checkPersistenceContextForCachedResults( naturalIds, session, lockOptions, results );
+		if ( !isEmpty( unresolvedIds ) ) {
+			loadEntitiesWithUnresolvedIds( unresolvedIds, session, lockOptions, results );
+		}
+
+		return results;
+	}
```

Comment:
I don't understand the purpose of this change. What's the reason for creating a `List` in the supertype and then asking the subtypes to add their results to it, instead of just asking the subtype for a `List` of results? Naively it looks like we just wind up creating two lists, but it's hard to tell what I'm missing from just looking at the diff.

Reply:
The loading of unreolved Ids is what is different between `MultiNaturalIdLoaderArrayParam` and `MultiNaturalIdLoaderInPredicate`, the rest of the logic (checking session) is common to all hence done in the base class.
I'm not sure I understand your concern here?

Reply:
The question is why doesn't `loadEntitiesWithUnresolvedIds()` just return a `List`, instead of receiving an empty mutable `List` as an argument?

Reply:
I guess maybe I wanted to avoid creating another List when I already have the one I need, don't really remember my reasoning for it. I'll change that.